adapters/http-base/src/main/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapter.java

```diff
@@ -856,31 +838,23 @@ public Future<Integer> getTimeUntilDisconnect(final TenantObject tenant, final I
      * The handler will close the message consumer and increment the metric for expired TTDs.
      *
      * @param ctx The context to retrieve cookies and the HTTP response from.
-     * @param messageConsumer The message consumer to receive a command. If {@code null}, no handler is added.
+     * @param responseReadyTracker The response ready tracker.
      * @param tenantId The tenant that the device belongs to.
      * @param deviceId The identifier of the device.
-     * @param currentSpan The <em>OpenTracing</em> Span used for tracking the processing of the request.
      */
     private void setTtdRequestConnectionCloseHandler(
             final RoutingContext ctx,
-            final CommandConsumer messageConsumer,
+            final ResponseReadyTracker responseReadyTracker,
             final String tenantId,
-            final String deviceId,
-            final Span currentSpan) {
+            final String deviceId) {
 
-        if (messageConsumer != null && !ctx.response().closed() && !ctx.response().ended()) {
+        if (!ctx.response().closed() && !ctx.response().ended()) {
             ctx.response().closeHandler(v -> {
-                log.debug("device [tenant: {}, device-id: {}] closed connection before response could be sent",
+                if (responseReadyTracker.cancel("device closed connection, stop waiting for command")) {
+                    log.debug("device [tenant: {}, device-id: {}] closed connection before response could be sent",
                         tenantId, deviceId);
-                currentSpan.log("device closed connection, stop waiting for command");
-                cancelCommandReceptionTimer(ctx);
```

Comment:
It looks to me that the log line and the `cancelCommandReceptionTimer` invocation should be done first in the closeHandler - before and independent of the `responseReadyTracker.cancel` invocation.

Reply:
I mean otherwise `cancelCommandReceptionTimer ` will only get invoked *after* `responseReady.complete()` is invoked.